security-jwt/src/main/java/io/micronaut/security/token/jwt/signature/jwks/JwksSignature.java

```diff
@@ -131,9 +131,27 @@ public Integer getCacheExpiration() {
 
     @Override
     @NonNull
-    public Optional<Boolean> isJwksCacheExpired() {
-        return (jwkSetCachedAt == null) ? Optional.empty() :
-                Optional.of(!new Date().before(jwkSetCachedAt));
+    public synchronized Optional<Boolean> isJwksCacheExpired() {
+        Date cachedAt = this.jwkSetCachedAt;
+        return isExpired(cachedAt);
+    }
+
+    @Override
+    public synchronized void clearJwksCache() {
+        this.jwkSet = null;
+        this.jwkSetCachedAt = null;
+    }
+
+    @Override
+    public synchronized boolean isJwksCachePresent() {
+        return this.jwkSet != null;
+    }
+
+    @Override
+    @NonNull
+    public synchronized Optional<List<String>> getJwksKeyIDs() {
+        JWKSet jwkSet = this.jwkSet;
+        return parseJwksKeyIDs(jwkSet);
```

Comment:
So while this does fix it I don't think it was the right fix as it adds *a ton* of potential contention between requests. Note that a few of these are used by `comparator` from `validate` which is in the hot path of requests.

Rather than a sledgehammer (synchronized) I would recommend a scalpel (store it in a local variable so it doesn't check the null then NPE a couple instructions later). In this case it is actually probably okay if it compares against slightly out of date data based on the intended usages in order to not block other requests.

Reply:
Agreed I don't see the need for synchronization here

Reply:
I think I take that back because we need synchronization because we have separate operations, check for null, get key set